### ISSUE TITLE
Empty derivationStack / Cannot read property 'observing' of undefined

### ISSUE THREAD

**Issue by sgoll** [mobxjs/mobx#328]
mobx 2.3.0 throws an uncaught TypeError: `Cannot read property 'observing' of undefined` an `reportObserved`. Here is the stack trace:

```
Uncaught TypeError: Cannot read property 'observing' of undefined
reportObserved @ mobx.js:943
ComputedValue.get @ mobx.js:699
(anonymous function) @ mobx.js:192
get @ mobx.js:2209
subscribe @ index.js:1573
subscribe @ index.js:1779
forEach @ ramda.js:1809
(anonymous function) @ ramda.js:429
(anonymous function) @ ramda.js:466
f1 @ ramda.js:446
forEach @ ramda.js:1809
(anonymous function) @ ramda.js:429
(anonymous function) @ ramda.js:466
f1 @ ramda.js:446
(anonymous function) @ ramda.js:296
(anonymous function) @ ramda.js:48
forEach @ ramda.js:1809
(anonymous function) @ ramda.js:429
(anonymous function) @ ramda.js:466
f1 @ ramda.js:446
(anonymous function) @ ramda.js:296
(anonymous function) @ ramda.js:48
(anonymous function) @ index.js:1783
transaction @ mobx.js:1111
handleConnect @ index.js:1762
EventEmitter.emit @ events.js:74
(anonymous function) @ index.js:1387
```

The flow (from bottom to top) is as follows:
- an event happens (WebSocket connect)
- a mobx observable `connected` is set to true within a transaction
- inside the same transaction, some nested structures are traversed (`subscribe` is called on them)
- for one of those structures, a mobx computed value is read (which derives from the observable `connected` which was set above)
- on reading that computed value, mobx throws the ecxeption

To be honest, I cannot imagine how this error can happen. The line in question that throws the exception is `var deps = derivationStack[derivationStack.length - 1].observing;` So it seems that for whatever reason the derivationStack is empty when it shouldn't be.

Any hints how I can trace this error or any idea what I could be doing wrong?

**Comment by mweststrate** [comment]
Hmm it should not be possible for this to happen indeed. Could you setup a reproduction somewhere?

**Comment by mweststrate** [comment]
BTW it didn't happen with 2.2.2?

Op wo 15 jun. 2016 12:02 schreef sgoll <notifications@github.com>:

**Comment by sgoll** [comment]
I just tested it with the older version: the same exception is thrown with 2.2.2. I'll try and come up with a minimal reproduction.

**Comment by sgoll** [comment]
The bug seems to be related to the following: in my code, I access the observable in question during the `becomeObserved` handler of another Atom.

I created this minimal example: https://gist.github.com/sgoll/ea9e1287bba65b02d7c148a241ecbbad

In this example, the exception is raised during the initial render. The stack trace looks like this:

```
reportObserved	@	mobx.js944
Atom.reportObserved	@	mobx.js:632
ObservableValue.get	@	mobx.js:2099
Object.defineProperty.get	@	mobx.js:1976
subscribe	@	test.js:20
(anonymous function)	@	test.js:10
addObserver	@	mobx.js:931
bindDependencies	@	mobx.js:858
trackDerivedFunction	@	mobx.js:831
Reaction.track	@	mobx.js:1015
reactiveRender	@	index.js:63
initialRender	@	index.js:57
ReactCompositeComponentMixin._renderValidatedComponentWithoutOwnerOrContext	@	ReactCompositeComponent.js:815
(…)
```

Should this be possible? Or am I doing something that is not allowed?

**Comment by sgoll** [comment]
FWIW, the error did not happen in mobx@<2.2.

**Comment by mweststrate** [comment]
Thanks for the nice reproduction! Fix has been released as 2.3.2
